src/components/RequestGroupComposition/InstrumentConfigForm.vue

```diff
@@ -0,0 +1,159 @@
+        v-model="instrumentConfig.optical_elements[opticalElementGroup.type]"
+        :field="opticalElementGroup.type"
+        :label="getFromObject(formConfig, ['instrumentConfig', opticalElementGroup.type, 'label'], opticalElementGroup.label)"
+        :desc="getFromObject(formConfig, ['instrumentConfig', opticalElementGroup.type, 'desc'], '')"
+        :hide="getFromObject(formConfig, ['instrumentConfig', opticalElementGroup.type, 'hide'], false)"
+        :options="opticalElementGroup.options"
+        lower-options
+        :tooltip-config="tooltipConfig"
+        :errors="{}"
+        @input="updateOpticalElement"
+      />
+    </div>
+    <div v-if="rotatorModeOptions.length > 0">
+      <custom-select
+        v-model="instrumentConfig.rotator_mode"
+        field="rotator_mode"
+        :label="getFromObject(formConfig, ['instrumentConfig', 'rotator_mode', 'label'], 'Rotator Mode')"
+        :desc="getFromObject(formConfig, ['instrumentConfig', 'rotator_mode', 'desc'], '')"
+        :hide="getFromObject(formConfig, ['instrumentConfig', 'rotator_mode', 'hide'], false)"
+        :tooltip-config="tooltipConfig"
+        :errors="errors.rotator_mode"
+        :options="rotatorModeOptions"
+        @input="update"
+      />
+      <!-- TODO: Validate required field values -->
+      <custom-field
+        v-for="field in requiredRotatorModeFields"
+        :key="field"
+        v-model="instrumentConfig.extra_params[field]"
+        :label="getFromObject(formConfig, ['instrumentConfig', field, 'label'], field)"
+        :desc="getFromObject(formConfig, ['instrumentConfig', field, 'desc'], '')"
+        :hide="getFromObject(formConfig, ['instrumentConfig', field, 'desc'], false)"
+        :tooltip-config="tooltipConfig"
+        :errors="null"
+        @input="updateInstrumentConfigExtraParam($event, field)"
+      />
+    </div>
+  </b-form>
+</template>
+<script>
+import { toRef } from '@vue/composition-api';
+
+import baseInstrumentConfig from '@/composables/baseInstrumentConfig.js';
+import CustomField from '@/components/RequestGroupComposition/CustomField.vue';
+import CustomSelect from '@/components/RequestGroupComposition/CustomSelect.vue';
+import { getFromObject, defaultTooltipConfig } from '@/util';
+
+export default {
+  name: 'InstrumentConfigForm',
+  components: {
+    CustomField,
+    CustomSelect
+  },
+  props: {
+    errors: {
+      type: Object,
+      required: true
+    },
+    availableInstruments: {
+      type: Object,
+      required: true
+    },
+    selectedInstrument: {
+      type: String,
+      required: true
+    },
+    instrumentConfig: {
+      type: Object,
+      required: true
+    },
+    show: {
+      type: Boolean
+    },
+    tooltipConfig: {
+      type: Object,
+      default: () => {
+        return defaultTooltipConfig;
+      }
+    },
+    formConfig: {
+      type: Object,
+      default: () => {
+        return {};
+      }
+    }
+  },
+  setup: function(props, context) {
```

Comment:
This setup hook is using the new Vue 3 Composition API. `baseInstrumentConfig` is a composition function, which is defined in `src/composables/baseInstrumentConfig.js`. Basically, any lifecycle hooks in that function are run in this component, and anything that is returned from the setup function is available in the rest of the components. A single component can use many composition functions, each of which could implement a different feature.

Reply:
This is slick!